File: CS/DeleteItemFromListBox/DeleteHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using System.Windows.Forms;
using System.Drawing;
using DevExpress.XtraEditors.Controls;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.ViewInfo;

namespace DeleteItemFromListBox
{
    public class DeleteHelper
    {
        ListBoxControl _list;
        int itemIndex = -1;
        RepositoryItemButtonEdit repository;
        EditorButton editorButton;
        int size = 18;
        ObjectState state;

        public DeleteHelper(ListBoxControl list)
        {
            _list = list;
            _list.MouseMove += _MouseMove;
            _list.DrawItem += _DrawItem;
            _list.MouseClick += _MouseClick;
            _list.MouseDown += _MouseDown;
            _list.ItemHeight = size + 1;
            repository = new RepositoryItemButtonEdit();
            repository.Buttons.Clear();
            editorButton = new EditorButton(ButtonPredefines.Close);
            repository.Buttons.Add(editorButton);
            repository.TextEditStyle = TextEditStyles.HideTextEditor;
            repository.BorderStyle = BorderStyles.NoBorder;
        }

        void _MouseDown(object sender, MouseEventArgs e)
        {
            if (itemIndex < 0)
                return;
            if (IsButton(e.Location))
            {
                state = ObjectState.Pressed;
                Redraw();
            }
        }

        void _MouseClick(object sender, MouseEventArgs e)
        {
            if (itemIndex < 0)
                return;
            if (IsButton(e.Location))
                _list.Items.RemoveAt(itemIndex);
        }

        private void DrawButton(ListBoxDrawItemEventArgs e)
        {
            Rectangle bounds = GetRectangle(e.Bounds);
            ButtonEditViewInfo buttonEditViewInfo = (ButtonEditViewInfo)repository.CreateViewInfo();
            buttonEditViewInfo.Bounds = bounds;
            buttonEditViewInfo.CalcViewInfo(e.Graphics);
            DevExpress.XtraEditors.Drawing.EditorButtonObjectInfoArgs buttonInfoByButton = buttonEditViewInfo.ButtonInfoByButton(editorButton);
            buttonInfoByButton.State = state;
            DevExpress.XtraEditors.Drawing.ControlGraphicsInfoArgs info = new DevExpress.XtraEditors.Drawing.ControlGraphicsInfoArgs(buttonEditViewInfo, e.Cache, bounds);
            repository.CreatePainter().Draw(info);
        }

        private static void DrawContent(ListBoxDrawItemEventArgs e)
        {
            e.Appearance.FillRectangle(e.Cache, e.Bounds);
            Rectangle textRect = Rectangle.Inflate(e.Bounds, -2, 0);
            e.Appearance.DrawString(e.Cache, e.Item.ToString(), textRect);
        }

        void _DrawItem(object sender, ListBoxDrawItemEventArgs e)
        {
            if (itemIndex < 0 || e.Index != itemIndex)
                return;
            DrawContent(e);
            DrawButton(e);
            e.Handled = true;
        }

        private void _MouseMove(object sender, MouseEventArgs e)
        {
            int newIndex = _list.IndexFromPoint(e.Location);
            if (itemIndex >= 0)
                Redraw();
            itemIndex = newIndex;
            if (itemIndex < 0)
                return;
            if (IsButton(e.Location))
                state = ObjectState.Hot;
            else
                state = ObjectState.Normal;
            Redraw();
        }

        Rectangle GetRectangle(Rectangle source)
        {
            Rectangle result = source;
            result.X = source.Right - size;
            result.Width = size;
            result.Height = size;
            return result;
        }

        void Redraw()
        {
            _list.Invalidate(_list.GetItemRectangle(itemIndex));
        }

        bool IsButton(Point p)
        {
            return GetRectangle(_list.GetItemRectangle(itemIndex)).Contains(p);
        }
    }
}

File: VB/DeleteItemFromListBox/DeleteHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports System.Windows.Forms
Imports System.Drawing
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.ViewInfo

Namespace DeleteItemFromListBox
	Public Class DeleteHelper
		Private _list As ListBoxControl
		Private itemIndex As Integer = -1
		Private repository As RepositoryItemButtonEdit
		Private editorButton As EditorButton
		Private size As Integer = 18
		Private state As ObjectState

		Public Sub New(ByVal list As ListBoxControl)
			_list = list
			AddHandler _list.MouseMove, AddressOf _MouseMove
			AddHandler _list.DrawItem, AddressOf _DrawItem
			AddHandler _list.MouseClick, AddressOf _MouseClick
			AddHandler _list.MouseDown, AddressOf _MouseDown
			_list.ItemHeight = size + 1
			repository = New RepositoryItemButtonEdit()
			repository.Buttons.Clear()
			editorButton = New EditorButton(ButtonPredefines.Close)
			repository.Buttons.Add(editorButton)
			repository.TextEditStyle = TextEditStyles.HideTextEditor
			repository.BorderStyle = BorderStyles.NoBorder
		End Sub

		Private Sub _MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
			If itemIndex < 0 Then
				Return
			End If
			If IsButton(e.Location) Then
				state = ObjectState.Pressed
				Redraw()
			End If
		End Sub

		Private Sub _MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
			If itemIndex < 0 Then
				Return
			End If
			If IsButton(e.Location) Then
				_list.Items.RemoveAt(itemIndex)
			End If
		End Sub

		Private Sub DrawButton(ByVal e As ListBoxDrawItemEventArgs)
			Dim bounds As Rectangle = GetRectangle(e.Bounds)
			Dim buttonEditViewInfo As ButtonEditViewInfo = CType(repository.CreateViewInfo(), ButtonEditViewInfo)
			buttonEditViewInfo.Bounds = bounds
			buttonEditViewInfo.CalcViewInfo(e.Graphics)
			Dim buttonInfoByButton As DevExpress.XtraEditors.Drawing.EditorButtonObjectInfoArgs = buttonEditViewInfo.ButtonInfoByButton(editorButton)
			buttonInfoByButton.State = state
			Dim info As New DevExpress.XtraEditors.Drawing.ControlGraphicsInfoArgs(buttonEditViewInfo, e.Cache, bounds)
			repository.CreatePainter().Draw(info)
		End Sub

		Private Shared Sub DrawContent(ByVal e As ListBoxDrawItemEventArgs)
			e.Appearance.FillRectangle(e.Cache, e.Bounds)
			Dim textRect As Rectangle = Rectangle.Inflate(e.Bounds, -2, 0)
			e.Appearance.DrawString(e.Cache, e.Item.ToString(), textRect)
		End Sub

		Private Sub _DrawItem(ByVal sender As Object, ByVal e As ListBoxDrawItemEventArgs)
			If itemIndex < 0 OrElse e.Index <> itemIndex Then
				Return
			End If
			DrawContent(e)
			DrawButton(e)
			e.Handled = True
		End Sub

		Private Sub _MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim newIndex As Integer = _list.IndexFromPoint(e.Location)
			If itemIndex >= 0 Then
				Redraw()
			End If
			itemIndex = newIndex
			If itemIndex < 0 Then
				Return
			End If
			If IsButton(e.Location) Then
				state = ObjectState.Hot
			Else
				state = ObjectState.Normal
			End If
			Redraw()
		End Sub

		Private Function GetRectangle(ByVal source As Rectangle) As Rectangle
			Dim result As Rectangle = source
			result.X = source.Right - size
			result.Width = size
			result.Height = size
			Return result
		End Function

		Private Sub Redraw()
			_list.Invalidate(_list.GetItemRectangle(itemIndex - 1))
			_list.Invalidate(_list.GetItemRectangle(itemIndex))
		End Sub

		Private Function IsButton(ByVal p As Point) As Boolean
			Return GetRectangle(_list.GetItemRectangle(itemIndex)).Contains(p)
		End Function
	End Class
End Namespace
